text/0097-log-data-model.md

```diff
@@ -0,0 +1,971 @@
+to display the short name in such UI elements.
+
+For example a dropdown list of severities that allows filtering log records by
+severities is likely to be more usable if it contains the short names of
+`SeverityNumber` (and thus has a limited upper bound of elements) compared to a
+dropdown list, which lists all distinct `SeverityText` values that are known to
+the system (which can be a large number of elements, often differing only in
+capitalization or abbreviated, e.g. "Info" vs "Information").
+
+#### Comparing Severity
+
+In the contexts where severity participates less-than / greater-than comparisons
+`SeverityNumber` field should be used. `SeverityNumber` can be compared to
+another `SeverityNumber` or to numbers in the 1..24 range (or to the
+corresponding short names).
+
+When severity is used in equality or inequality comparisons (for example in
+filters in the UIs) the recommendation is to attempt to use both `SeverityText`
+and short name of `SeverityNumber` to perform matches. For example if we have a
+record with `SeverityText` field equal to "Informational" and `SeverityNumber`
+field equal to INFO then it may be preferable from the user experience
+perspective to ensure that **severity="Informational"** and **severity="INFO"**
+conditions both to are TRUE for that record.
+
+### Field: `ShortName`
+
+Type: string.
+
+Description: Short event identifier that does not contain varying parts.
+`ShortName` describes what happened (e.g. "ProcessStarted"). Recommended to be
+no longer than 50 characters. Optional. Not guaranteed to be unique in any way.
+Typically used for filtering and grouping purposes in backends.
+
+### Field: `Body`
+
+Type: any.
+
+Description: A value containing the body of the log record (see the description
+of `any` type above). Can be for example a human-readable string message
+(including multi-line) describing the event in a free form or it can be a
+structured data composed of arrays and maps of other values. Can vary for each
+occurrence of the event coming from the same source.
+
+### Field: `Resource`
+
+Type: key/value pair list.
+
+Description: Describes the source of the log, aka
+[resource](https://github.com/open-telemetry/opentelemetry-specification/blob/master/specification/overview.md#resources).
+"value" of each pair is of `any` type. Multiple occurrences of events coming
+from the same event source can happen across time and they all have the same
+value of `Resource`. Can contain for example information about the application
+that emits the record or about the infrastructure where the application runs.
+Data formats that represent this data model may be designed in a manner that
+allows the `Resource` field to be recorded only once per batch of log records
+that come from the same source. SHOULD follow OpenTelemetry
+[semantic conventions for Resources](https://github.com/open-telemetry/opentelemetry-specification/tree/master/specification/resource/semantic_conventions).
```

Comment:
you may sometimes have a log even that relates to a resource from another source. For example, imagine if we were parsing events from Kubernetes, which _talks about_ many different containers and potentially other resources.  

I think we should perhaps consider removing the text that talks about how "source" and "Resource" relates and that for the same source, you do not need to repeat the "resource" for each item in a batch. 
 It's my experience that in most cases, compression will minimize the cost of repetition.

Reply:
I understand. What you say makes a lot of sense.

Firstly, the definition of the Resource in this proposal is consistent with how it is used for Traces and Metrics in OpenTelemetry. This is simply re-iteration of what already is part of the Traces and Metrics spec. I believe it is highly important to keep the consistency between 3 signals unless there are good reasons to diverge.

On the other hand, what you describe I believe is applicable not just to logs but to other telemetry signals as well. For instance an application A may report a metric such as "request count" and it may be the number of requests that application A made to application B. Thus, this requires that the metric references multiple resources.

I think we need a general capability for telemetry signals to reference more than one resource. I still think that the current definition of the Resource is not wrong. It indicates where the signal was generated and it is likely that in the majority of use cases there is never a second resource to reference. However, again, I do believe you are right and we need a way to reference the second (or third, etc) resources.

I believe it is an extension to OpenTelemetry's undestanding of resources and needs to be done separately, outside of this OTEP, because it needs to apply universally to all 3 signals. It is a significant change for OpenTelemetry spec and requires its own OTEP. If that new OTEP gets accepted we can modify the Log Data Model correspondingly.

There are likely many ways to record multiple resources in one record, hopefully also in  a way that is backwards compatible with the current spec. For example: one approach would be that additional resources may be referenced in attributes, e.g. an attribute may be defined to contain a "Resource" as a value, e.g. (using made-up yaml syntax for data):

```yaml
logrecord:
  timestamp: 1588372122
  resource:
     service.name: A
  body: Incoming request
  attributes:
    http.path: /request/path
    request.origin: # semantic conventions define that "request.origin" is a resource.
      service.name: B
```

Reply:
It is mandatory to have `Resource` for all signals to ensure correlation between signals. The `Resource` describes the source of the telemetry.

1. If the log entry is generated by a container, then propagated by any other k8s component, the resource should describe the container that generated the log entry.
2. If the log entry is generated by a k8s component and it talks about a container, the `Resource` should describe the k8s component that generated the log entry. It is up to the backend or other analyzing tools to parse the log entry and determine that the entry may reference other `Resource` but still the source of the log entry is the k8s component that generated the log.

Reply:
Captured an issue to discuss separately: https://github.com/open-telemetry/opentelemetry-specification/issues/579